src/Symfony/Component/HttpClient/CurlHttpClient.php

```diff
@@ -0,0 +1,314 @@
+            return;
+        }
+
+        // Keep a dummy "onPush" reference to work around a refcount bug in PHP
+        curl_multi_setopt($mh, CURLMOPT_PUSHFUNCTION, $multi->onPush = static function ($parent, $pushed, array $rawHeaders) use ($multi) {
+            $headers = [];
+
+            foreach ($rawHeaders as $h) {
+                if (false !== $i = strpos($h, ':', 1)) {
+                    $headers[substr($h, 0, $i)] = substr($h, 1 + $i);
+                }
+            }
+
+            if ('GET' !== $headers[':method'] || isset($headers['range'])) {
+                return CURL_PUSH_DENY;
+            }
+
+            $url = $headers[':scheme'].'://'.$headers[':authority'];
+
+            // curl doesn't validate the pushed ":authority" header, but this is a MUST in the HTTP/2 RFC
+            // doing full validation requires evaluating cert's alt names, origin frames and even the alt-svc
+            // let's keep it simple and do URL-based validation for now
+            if (0 !== strpos(curl_getinfo($parent, CURLINFO_EFFECTIVE_URL), $url.'/')) {
+                return CURL_PUSH_DENY;
+            }
+
+            $multi->pushedResponses[$url.$headers[':path']] = [
+                new CurlResponse($multi, $pushed),
+                [
+                    $headers['authorization'] ?? null,
+                    $headers['cookie'] ?? null,
+                    $headers['x-requested-with'] ?? null,
+                    null,
+                ],
+            ];
+
+            return CURL_PUSH_OK;
+        });
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function request(string $method, string $url, iterable $options = []): ResponseInterface
+    {
+        [$url, $options] = self::prepareRequest($method, $url, $options, $this->defaultOptions);
+        $scheme = $url['scheme'];
+        $authority = $url['authority'];
+        $url = implode('', $url);
+
+        if ([$pushedResponse, $pushedHeaders] = $this->multi->pushedResponses[$url] ?? null) {
+            unset($this->multi->pushedResponses[$url]);
+            // Accept pushed responses only if their headers related to authentication match the request
+            $expectedHeaders = [
+                $options['headers']['authorization'] ?? null,
+                $options['headers']['cookie'] ?? null,
+                $options['headers']['x-requested-with'] ?? null,
+                $options['headers']['range'] ?? null,
+            ];
+
+            if ('GET' === $method && !$options['body'] && $expectedHeaders === $pushedHeaders) {
+                // Reinitialize the pushed response with request's options
+                return $pushedResponse->__construct($this->multi, $url, $options);
+            }
+        }
+
+        $curlopts = [
+            CURLOPT_URL => $url,
+            CURLOPT_DNS_USE_GLOBAL_CACHE => false,
+            CURLOPT_USERAGENT => 'Symfony HttpClient/Curl',
```

Comment:
This cannot be passed as an option ?

Reply:
Not directly as an option, but it can be defined as a header, and the header will have precedence.